### PR TITLE
Add chainerx.elu

### PR DESCRIPTION
**Pull request by aksub99**
Add `chainerx.elu`.

### CODE REVIEW

**Comment by aksub99** [comment]
@niboshi What could be the reason for the failing backward tests?

File: chainerx_cc/chainerx/routines/CMakeLists.txt [hunk 1]
```diff
@@ -1,5 +1,9 @@
 add_library(chainerx_routines STATIC
+<<<<<<< HEAD
     arithmetic.cc
+=======
+    activation.cc
+>>>>>>> Add chainerx.elu
     binary.cc
```
Comment (emcastillo) [inline chainerx_cc/chainerx/routines/CMakeLists.txt]: There is a syntax error in cmake here

Comment (emcastillo) [reply chainerx_cc/chainerx/routines/CMakeLists.txt]: The build is failing due to an incorrect CMakeLists,
Please fix it :)

Comment (aksub99) [reply chainerx_cc/chainerx/routines/CMakeLists.txt]: Thanks @emcastillo !

**Comment by aksub99** [comment]
@niboshi Could you please review?

**Comment by niboshi** [comment]
@aksub99 @emcastillo Thank you for the PR & comments!

File: chainerx_cc/chainerx/routines/activation.cc [hunk 2]
```diff
@@ -0,0 +1,21 @@
+
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/activation.cc]: Extra blank line?

File: chainerx_cc/chainerx/routines/activation.cc [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+#include "chainerx/array.h"
+#include "chainerx/constant.h"
+#include "chainerx/device.h"
+#include "chainerx/dtype.h"
+#include "chainerx/kernels/math.h"
+#include "chainerx/routines/activation.h"
+#include "chainerx/routines/creation.h"
+#include "chainerx/routines/explog.h"
+#include "chainerx/routines/indexing.h"
+#include "chainerx/scalar.h"
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/activation.cc]: * Move ` #include "chainerx/routines/activation.h" ` to the first line of the file.
* It seems some includes are not necessary (constant.h, dtype.h, kernels/math.h, <vector>)

File: chainerx_cc/chainerx/routines/activation.cc [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace chainerx {
+Array Elu(const Array& x, Scalar alpha) {
+    Array zero = ZerosLike(x, x.device());
+    return Where(x > zero, x, alpha * Expm1(x));
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/activation.cc]: `x > zero` could be replaced with `x > 0` if `operator>` supports scalars.
Could you write a TODO?

File: chainerx_cc/chainerx/routines/activation.h [hunk 5]
```diff
@@ -0,0 +1,14 @@
+#pragma once
+
+#include <cstdint>
+
+#include <nonstd/optional.hpp>
+
+#include "chainerx/array.h"
+#include "chainerx/constant.h"
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/activation.h]: * Some includes are unnecessary. (cstdint, optional, constant)
* Add scalar.h.

File: tests/chainerx_tests/unit_tests/routines_tests/test_activation.py [hunk 6]
```diff
@@ -75,6 +75,37 @@ def forward_xp(self, inputs, xp):
...
+        'shape': [(2, 3)],
+        'in_dtypes,out_dtype': _in_out_float_dtypes_math_functions,
+        'input': [0, float('inf'), -float('inf'), float('nan')],
+        'skip_backward_test': [True],
+        'skip_double_backward_test': [True],
+    })
+))
+class TestElu(UnaryMathTestBase, op_utils.NumpyOpTest):
+
+    alpha = 2.0
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_activation.py]: Please parameterize `alpha`, including a negative value, zero and "unspecified".

**Comment by aksub99** [comment]
@niboshi Sorry for the delay in addressing your comments. I have made all the suggested fixes. The tests are failing because of mismatch in `dtypes` of outputs and gradients. Could the gradients be having a different `dtype` because of `alpha`? If so, how can I change the type of  `alpha` to be the same as the output `dtype`(Since `alpha` might be a tuple)?

File: tests/chainerx_tests/unit_tests/routines_tests/test_activation.py [hunk 7]
```diff
@@ -72,6 +73,44 @@ def forward_xp(self, inputs, xp):
...
+        else:
+            self.alpha = self.alpha_range
+
+    def func(self, xp, a):
+        if xp is numpy:
+            y = a.copy()
+            negzero_indices = y <= 0
+            y[negzero_indices] = self.alpha * numpy.expm1(y[negzero_indices])
+            return y
+        return xp.elu(a, self.alpha)
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_activation.py]: Thank you for the fix.
Please also test "unspecified" alpha.
You could implement similarly to existing tests like [test_connection.py](https://github.com/chainer/chainer/blob/v7.0.0a1/tests/chainerx_tests/unit_tests/routines_tests/test_connection.py#L427) (define `Unspecified` class and parameterize it as `alpha_range`).

**Comment by niboshi** [comment]
Please check Travis failure, too.

File: tests/chainerx_tests/unit_tests/routines_tests/test_activation.py (hunk 7 above)
Comment (aksub99) [reply tests/chainerx_tests/unit_tests/routines_tests/test_activation.py]: Sure. Will modify the parametrizations accordingly.

**Comment by aksub99** [comment]
@niboshi I have parametrized `alpha` with Unspecified.
I'm unsure of how to fix the travis failures. I thought the mismatch in dtypes might be because of alpha's dtype but I guess I was mistaken.

File: tests/chainerx_tests/unit_tests/routines_tests/test_activation.py [hunk 8]
```diff
@@ -72,6 +78,49 @@ def forward_xp(self, inputs, xp):
...
+        'alpha_range': [(-2.0, 0.0), 0.0, (0.0, 2.0), Unspecified],
+    })
+))
+class TestElu(UnaryMathTestBase, op_utils.NumpyOpTest):
+
+    def setup(self):
+        if isinstance(self.alpha_range, tuple):
+            l, u = self.alpha_range
+            self.alpha = random.uniform(l, u)
+        elif self.alpha_range in (None, Unspecified):
```
Comment (kshitij12345) [inline tests/chainerx_tests/unit_tests/routines_tests/test_activation.py]: Is `Unspecified` necessary?
With 
`'alpha_range': [(-2.0, 0.0), (0.0, 2.0),  0.0, 1.0]`

We will have
```Python
if isinstance(self.alpha_range, tuple):
  l, u = self.alpha_range
  self.alpha = random.uniform(l, u)
else:
  self.alpha = self.alpha_range
```
Which will have the same behavior as it does now.

**Comment by aksub99** [comment]
@niboshi I have fixed the travis failures and rebased. Please review:)

**Comment by niboshi** [comment]
Thanks!

File: tests/chainerx_tests/unit_tests/routines_tests/test_activation.py [hunk 9]
```diff
@@ -144,6 +152,60 @@ def func(self, xp, a):
...
+    })
+))
+class TestElu(UnaryMathTestBase, op_utils.NumpyOpTest):
+
+    def setup(self):
+        in_dtype, = self.in_dtypes
+        if isinstance(self.alpha_range, tuple):
+            l, u = self.alpha_range
+            self.alpha = random.uniform(l, u)
+        elif self.alpha_range in (None, Unspecified):
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_activation.py]: Can you remove `None`?

File: tests/chainerx_tests/unit_tests/routines_tests/test_activation.py [hunk 10]
```diff
@@ -144,6 +152,60 @@ def func(self, xp, a):
...
+            self.check_double_backward_options.update(
+                {'rtol': 1e-2, 'atol': 1e-2})
+
+    def func(self, xp, a):
+        if xp is numpy:
+            y = a.copy()
+            negzero_indices = y <= 0
+            y[negzero_indices] = self.alpha * numpy.expm1(y[negzero_indices])
+            return y
+        elif self.alpha_range in (None, Unspecified):
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_activation.py]: ditto.

File: chainerx_cc/chainerx/python/routines.cc [hunk 11]
```diff
@@ -625,12 +625,13 @@ void InitChainerxManipulation(pybind11::module& m) {
 void InitChainerxActivation(pybind11::module& m) {
     m.def("clipped_relu", [](const ArrayBodyPtr& x, Scalar z) { return MoveArrayBody(ClippedRelu(Array{x}, z)); }, "x"_a, "z"_a = 20.0);
     m.def("crelu", [](const ArrayBodyPtr& x, int8_t axis) { return MoveArrayBody(CRelu(Array{x}, axis)); }, "x"_a, "axis"_a = 1);
+    m.def("elu", [](const ArrayBodyPtr& x, float_t alpha) { return MoveArrayBody(Elu(Array{x}, alpha)); }, "x"_a, "alpha"_a = 1.0);
```
Comment (niboshi) [inline chainerx_cc/chainerx/python/routines.cc]: `float_t` -> `double`?

File: chainerx_cc/chainerx/routines/activation.cc [hunk 12]
```diff
@@ -38,6 +39,14 @@ Array CRelu(const Array& x, int8_t axis) {
     return Relu(concat);
 }
 
+Array Elu(const Array& x, float_t alpha) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/activation.cc]: ditto.

File: chainerx_cc/chainerx/routines/activation.h [hunk 13]
```diff
@@ -9,6 +9,8 @@ Array ClippedRelu(const Array& x, Scalar z);
 
 Array CRelu(const Array& x, int8_t axis);
 
+Array Elu(const Array& x, float_t alpha = 1.0);
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/activation.h]: ditto.

**Comment by aksub99** [comment]
@niboshi I have made the required changes.

**Comment by niboshi** [comment]
Thanks.
Jenkins, test this please

**Comment by aksub99** [comment]
The test failures seem to be caused by `group_normalization` which is unrelated to this PR. Am I correct?

**Comment by niboshi** [comment]
@aksub99 
Yes, you're right.
Thank you for the PR!

**Pull request merged**